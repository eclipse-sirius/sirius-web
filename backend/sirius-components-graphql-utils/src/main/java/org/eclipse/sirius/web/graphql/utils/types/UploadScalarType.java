/*******************************************************************************
 * Copyright (c) 2019, 2020 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
package org.eclipse.sirius.web.graphql.utils.types;

import graphql.schema.GraphQLScalarType;

/**
 * Custom {@link GraphQLScalarType} dedicated to Upload.
 *
 * @author SMonnier
 */
public class UploadScalarType {

    public static final String UPLOAD_TYPE = "Upload"; //$NON-NLS-1$

    // @formatter:off
    public static final GraphQLScalarType INSTANCE = GraphQLScalarType.newScalar()
            .name(UPLOAD_TYPE)
            .description("Upload type used for the file upload") //$NON-NLS-1$
            .coercing(new UploadScalarCoercing())
            .build();
    // @formatter:on
}
