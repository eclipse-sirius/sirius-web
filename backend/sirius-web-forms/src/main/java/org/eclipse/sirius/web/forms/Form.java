/*******************************************************************************
 * Copyright (c) 2019, 2020 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
package org.eclipse.sirius.web.forms;

import java.text.MessageFormat;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

import org.eclipse.sirius.web.annotations.Immutable;
import org.eclipse.sirius.web.annotations.graphql.GraphQLField;
import org.eclipse.sirius.web.annotations.graphql.GraphQLID;
import org.eclipse.sirius.web.annotations.graphql.GraphQLNonNull;
import org.eclipse.sirius.web.annotations.graphql.GraphQLObjectType;
import org.eclipse.sirius.web.representations.IRepresentation;
import org.eclipse.sirius.web.representations.ISemanticRepresentation;

/**
 * Root concept of the form representation.
 *
 * @author sbegaudeau
 * @author hmarchadour
 */
@Immutable
@GraphQLObjectType
public final class Form implements IRepresentation, ISemanticRepresentation {

    public static final String KIND = "Form"; //$NON-NLS-1$

    private String id;

    private String kind;

    private String label;

    private String targetObjectId;

    private UUID descriptionId;

    private List<Page> pages;

    private Form() {
        // Prevent instantiation
    }

    @Override
    @GraphQLID
    @GraphQLField
    @GraphQLNonNull
    public String getId() {
        return this.id;
    }

    @Override
    @GraphQLField
    @GraphQLNonNull
    public String getKind() {
        return this.kind;
    }

    @Override
    @GraphQLNonNull
    @GraphQLField
    public String getLabel() {
        return this.label;
    }

    @Override
    @GraphQLID
    @GraphQLField
    @GraphQLNonNull
    public String getTargetObjectId() {
        return this.targetObjectId;
    }

    @Override
    @GraphQLID
    @GraphQLField
    @GraphQLNonNull
    public UUID getDescriptionId() {
        return this.descriptionId;
    }

    @GraphQLNonNull
    @GraphQLField
    public List<@GraphQLNonNull Page> getPages() {
        return this.pages;
    }

    public static Builder newForm(String id) {
        return new Builder(id);
    }

    @Override
    public String toString() {
        String pattern = "{0} '{'id: {1}, label: {2}, targetObjectId: {3}, descriptionId: {4} pageCount: {5}'}'"; //$NON-NLS-1$
        return MessageFormat.format(pattern, this.getClass().getSimpleName(), this.id, this.label, this.targetObjectId, this.descriptionId, this.pages.size());
    }

    /**
     * The builder used to create the form.
     *
     * @author sbegaudeau
     */
    @SuppressWarnings("checkstyle:HiddenField")
    public static final class Builder {
        private String id;

        private String kind = KIND;

        private String label;

        private String targetObjectId;

        private UUID descriptionId;

        private List<Page> pages;

        private Builder(String id) {
            this.id = Objects.requireNonNull(id);
        }

        public Builder label(String label) {
            this.label = Objects.requireNonNull(label);
            return this;
        }

        public Builder targetObjectId(String targetObjectId) {
            this.targetObjectId = Objects.requireNonNull(targetObjectId);
            return this;
        }

        public Builder descriptionId(UUID descriptionId) {
            this.descriptionId = Objects.requireNonNull(descriptionId);
            return this;
        }

        public Builder pages(List<Page> pages) {
            this.pages = pages;
            return this;
        }

        public Form build() {
            Form form = new Form();
            form.id = Objects.requireNonNull(this.id);
            form.kind = Objects.requireNonNull(this.kind);
            form.label = Objects.requireNonNull(this.label);
            form.targetObjectId = Objects.requireNonNull(this.targetObjectId);
            form.descriptionId = Objects.requireNonNull(this.descriptionId);
            form.pages = Objects.requireNonNull(this.pages);
            return form;
        }
    }
}
