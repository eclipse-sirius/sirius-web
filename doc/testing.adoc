# Testing

This document will describe how we should try to write and maintain tests in Sirius Web.

IMPORTANT: TLDR: Use backend integration tests if you don’t have large and complex pieces of frontend code to test that involve the user interface (diagram interactions, diagram layout, form description editor drag and drop, etc).
They can easily be configured to test very complex use cases involving domains, views, custom services, etc.
They are very fast to execute and are very stable.


## What kind of tests do we have?

We have quite a large scope of tests in Sirius Web with:

- Backend unit tests
- Backend integration tests
- Frontend unit tests
- End to end tests
- Manual tests

## What kind of tests should be written?

The kind of tests that should be written depends on the work that has been done and which should be tested and the goal of the tests.

### Backend unit tests

This kind of tests should be used if you have some very complex algorithm in one specific method.
Given the kind of application that we are developing, this kind of tests should be quite rare.
While we had the backend layout algorithm in the past which represented quite a complex set of algorithms with some geometric computation, this is not really the case anymore.
Since most of the complexity we have to deal with is the integration between components, they are not that useful for us today.

### Frontend unit tests

Just like their backend counterparts, such tests should be written for complex pieces of algorithm on the frontend.
Some of our React component can easily have a complex lifecycle using complex DOM interactions and synchronisation steps with the remote server for example.
They could be tested using this approach but most of the time the experience has shown that this may require setting up some complex mocks of both the DOM and the backend.

Most of the complexity in our frontend code comes from the synchronisation of state between user interactions and the lifecycle of the backend.
As a result, most of the code that needs to be tested in the frontend could just as well be tested by end to end tests.

IMPORTANT: The more mocks are necessary to run unit tests, the further the tests are from the reality.
If you need to mock some parts of the database (with mocks of IRepository for backend tests), network interactions like HTTP requests / WebSocket messages (with mocks of useQuery() or useMutation()) or the DOM (jdom), you probably need integration tests instead.


### Backend integration tests

Those tests are to be considered if most of the logic to be tested is located on the backend and if the frontend part of the tested feature is not significant.

They allow you to test complex state of backend business logic.
It is possible to initialize complex state easily with snapshots of the database or custom services to initialize the editing context with domains or view based representation descriptions.
You can write tests to validate the complex steps of the lifecycle of the representations (create a representation, validate some initial state, trigger some tool, validate the impact of said tool).
We can even contribute test-specific services to mock customization provided by specifiers.

If you don’t have complex pieces of frontend code to test, they are the most efficient way to test Sirius Web applications.


### End to end tests with Cypress

End to end tests are a great way to manipulate the application like the end user.
They give you the ability to test features involving complex user interactions and frontend / backend exchanges.
For example, there is nothing like end to end tests to validate some frontend interactions in diagrams.

While they may seem perfect at first, they also have some sizable drawbacks.
End to end tests are very costly to setup.
Since we are limited to what the user interface and the HTTP API can do, it’s very complex to perform tests with complex initial state.
Testing new features of the view DSL for example, require the creation of a studio project and then an instance project and that’s both cumbersome, long and subject to errors.
 
It is also very difficult to test invalid use cases since both the user interface and the HTTP API prevent end users from creating issues.

End to end tests are also the most complex tests to write and maintain.
Since we have to interact with the user interface like end users, we have to write every single step taken by the end user even if they are not relevant for the feature to be tested.

These tests can easily be flaky since we have to carefully detect user interface states.
Countless tests have been invalid in the past because of a failure to detect that a server response had not been received before trying to perform an action.
For example, to create a representation, you need to wait for the response of the server in the new representation dialog otherwise the list of available representation is empty.


### Manual tests

This is THE perfect test environment since you will end up testing things using real human beings.
But since we are testing with humans, there are a couple things to consider.
Manual tests should be used for two main use cases.

Either to ensure the non regression of use cases that are so complex that we did not succeed in writing automated tests.
For example concurrent tests with multiple users.
Or to leverage the ingenuity of real human beings to find bugs instead of preventing regressions.

#### Using human beings to prevent regressions

Most of the time, preventing regressions involve starting from a specific state, executing a specific set of instructions and validating a specific end result.
This can and should be automated instead since:

- They can be complex to maintain since we won't easily know which tests should be updated after a release (other kinds of tests tends to break quickly after an incompatible change)
- They are boring to execute as an end user
- It’s a shame not to perform such specific test in the continuous integration process
- Human being are very good at missing small and boring details

As such using human for non regression is not really efficient. Let’s be more careful with the contribution of non regression tests when pull results are submitted.

#### Using humans to discover new bugs

Instead it is better to use human beings to use the application to find new bugs.
By giving them some freedom in the way each test should be executed, we can easily end up with different testers using different paths.
In order to discover bugs, the difference in the execution path can sometime be as subtle as clicking on menu items in a different order.
Those manual tests shouldn't be too complex or too long to not frustrate testers. 

## Decision

Here are a couple of directions in which we will move with regards to tests:

- If the vast majority of the behavior being tested is performed by the backend and if the frontend is not very complicated, we will default to backend integration tests
- If a sizable part of the behavior being tested is performed in the frontend, we will default to end to end tests or unit tests
- If a manual test can be automated, we will remove them and use an automated test instead
- End to end tests are very time consuming, we will try to restrain ourselves from writing this kind of test if backend integration tests can do the same job
- Flakky end to end tests are eroding our trust in the build by making everyone fine with red builds.
They will be removed and issues will be created to either restore them if the feature tested had to be tested using an end to end test or to convert them to backend integration tests.
