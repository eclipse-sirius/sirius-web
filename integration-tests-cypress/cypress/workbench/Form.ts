/*******************************************************************************
 * Copyright (c) 2024, 2025 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
export class Form {
  public getForm(): Cypress.Chainable<JQuery<HTMLElement>> {
    return cy.get(`[data-representation-kind="form"]`);
  }

  public getWidget(widgetLabel: string): Cypress.Chainable<JQuery<HTMLElement>> {
    return this.getForm().findByTestId(widgetLabel);
  }

  public getWidgetElement(widgetLabel: string, elementLabel: string): Cypress.Chainable<JQuery<HTMLElement>> {
    return this.getWidget(widgetLabel).findByTestId(elementLabel);
  }
}
