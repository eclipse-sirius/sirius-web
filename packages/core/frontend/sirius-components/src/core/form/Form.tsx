/*******************************************************************************
 * Copyright (c) 2019, 2022 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
import { makeStyles } from '@material-ui/core/styles';
import PropTypes from 'prop-types';
import React from 'react';

const useFormStyles = makeStyles((theme) => ({
  form: {
    display: 'flex',
    flexDirection: 'column',
    paddingTop: theme.spacing(1),
    paddingLeft: theme.spacing(2),
    paddingRight: theme.spacing(2),
    '& > *': {
      marginBottom: theme.spacing(2),
    },
  },
}));

const propTypes = {
  children: PropTypes.node.isRequired,
};
export const Form = ({ children, ...props }) => {
  const classes = useFormStyles();
  return (
    <form {...props} className={classes.form}>
      {children}
    </form>
  );
};
Form.propTypes = propTypes;
