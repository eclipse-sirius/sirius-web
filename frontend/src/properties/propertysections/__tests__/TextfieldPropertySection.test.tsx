/*******************************************************************************
 * Copyright (c) 2022 Obeo.
 * This program and the accompanying materials
 * are made available under the terms of the Eclipse Public License v2.0
 * which accompanies this distribution, and is available at
 * https://www.eclipse.org/legal/epl-2.0/
 *
 * SPDX-License-Identifier: EPL-2.0
 *
 * Contributors:
 *     Obeo - initial API and implementation
 *******************************************************************************/
import { MockedProvider, MockedResponse } from '@apollo/client/testing';
import { act, render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { Textfield } from 'form/Form.types';
import {
  editTextfieldMutation,
  TextfieldPropertySection,
  updateWidgetFocusMutation,
} from 'properties/propertysections/TextfieldPropertySection';
import {
  GQLEditTextfieldMutationData,
  GQLEditTextfieldMutationVariables,
  GQLEditTextfieldSuccessPayload,
  GQLErrorPayload,
  GQLUpdateWidgetFocusMutationData,
  GQLUpdateWidgetFocusMutationVariables,
  GQLUpdateWidgetFocusSuccessPayload,
} from 'properties/propertysections/TextfieldPropertySection.types';
import React from 'react';

jest.mock('uuid', () => ({ v4: () => '48be95fc-3422-45d3-b1f9-d590e847e9e1' }));

const defaultTextField: Textfield = {
  __typename: 'Textfield',
  id: 'textfieldId',
  label: 'Name:',
  stringValue: 'Composite Processor',
  diagnostics: [],
  style: null,
};

const textFieldWithStyle: Textfield = {
  __typename: 'Textfield',
  id: 'textfieldId',
  label: 'Name:',
  stringValue: 'Composite Processor',
  diagnostics: [],
  style: {
    backgroundColor: '#de1000',
    foregroundColor: '#fbb800',
  },
};

const textFieldWithEmptyStyle: Textfield = {
  __typename: 'Textfield',
  id: 'textfieldId',
  label: 'Name:',
  stringValue: 'Composite Processor',
  diagnostics: [],
  style: {
    backgroundColor: '',
    foregroundColor: '',
  },
};

const editTextfieldVariables: GQLEditTextfieldMutationVariables = {
  input: {
    id: '48be95fc-3422-45d3-b1f9-d590e847e9e1',
    editingContextId: 'editingContextId',
    representationId: 'formId',
    textfieldId: 'textfieldId',
    newValue: 'Main Composite Processor',
  },
};
const editTextfieldSuccessPayload: GQLEditTextfieldSuccessPayload = { __typename: 'EditTextfieldSuccessPayload' };
const editTextfieldSuccessData: GQLEditTextfieldMutationData = { editTextfield: editTextfieldSuccessPayload };

const updateWidgetFocusVariables: GQLUpdateWidgetFocusMutationVariables = {
  input: {
    id: '48be95fc-3422-45d3-b1f9-d590e847e9e1',
    editingContextId: 'editingContextId',
    representationId: 'formId',
    widgetId: 'textfieldId',
    selected: true,
  },
};
const updateWidgetFocusSuccessPayload: GQLUpdateWidgetFocusSuccessPayload = {
  __typename: 'UpdateWidgetFocusSuccessPayload',
};
const updateWidgetFocusSuccessData: GQLUpdateWidgetFocusMutationData = {
  updateWidgetFocus: updateWidgetFocusSuccessPayload,
};

const updateWidgetFocusErrorPayload: GQLErrorPayload = {
  __typename: 'ErrorPayload',
  message: 'An error has occurred, please refresh the page',
};
const updateWidgetFocusErrorData: GQLUpdateWidgetFocusMutationData = {
  updateWidgetFocus: updateWidgetFocusErrorPayload,
};

test('should render the textfield', () => {
  const { container } = render(
    <MockedProvider>
      <TextfieldPropertySection
        editingContextId="editingContextId"
        formId="formId"
        widget={defaultTextField}
        subscribers={[]}
        readOnly={false}
      />
    </MockedProvider>
  );
  expect(container).toMatchSnapshot();
});

test('should render a readOnly textfield', () => {
  const { container } = render(
    <MockedProvider>
      <TextfieldPropertySection
        editingContextId="editingContextId"
        formId="formId"
        widget={defaultTextField}
        subscribers={[]}
        readOnly
      />
    </MockedProvider>
  );
  expect(container).toMatchSnapshot();
});

test('should display the edited value', async () => {
  let updateWidgetFocusCalled = false;
  const updateWidgetFocusSuccessMock: MockedResponse<Record<string, any>> = {
    request: {
      query: updateWidgetFocusMutation,
      variables: updateWidgetFocusVariables,
    },
    result: () => {
      updateWidgetFocusCalled = true;
      return { data: updateWidgetFocusSuccessData };
    },
  };

  let editTextfieldCalled = false;
  const editTextfieldSuccessMock: MockedResponse<Record<string, any>> = {
    request: {
      query: editTextfieldMutation,
      variables: editTextfieldVariables,
    },
    result: () => {
      editTextfieldCalled = true;
      return { data: editTextfieldSuccessData };
    },
  };

  const mocks = [updateWidgetFocusSuccessMock, editTextfieldSuccessMock];
  const { container } = render(
    <MockedProvider mocks={mocks}>
      <TextfieldPropertySection
        editingContextId="editingContextId"
        formId="formId"
        widget={defaultTextField}
        subscribers={[]}
        readOnly={false}
      />
    </MockedProvider>
  );
  expect(container).toMatchSnapshot();

  const element: HTMLInputElement = screen.getByTestId('Name:').querySelector('input');
  expect(element.value).toBe('Composite Processor');

  await act(async () => {
    userEvent.click(element);
    await new Promise((resolve) => setTimeout(resolve, 0));

    await waitFor(() => {
      expect(updateWidgetFocusCalled).toBeTruthy();
      expect(element.value).toBe('Composite Processor');
      expect(container).toMatchSnapshot();
    });

    userEvent.clear(element);
    userEvent.type(element, 'Main Composite Processor{enter}');
    await new Promise((resolve) => setTimeout(resolve, 0));

    await waitFor(() => {
      expect(editTextfieldCalled).toBeTruthy();
      expect(element.value).toBe('Main Composite Processor');
      expect(container).toMatchSnapshot();
    });
  });
});

test('should display the error received', async () => {
  let updateWidgetFocusCalled = false;
  const updateWidgetFocusErrorMock: MockedResponse<Record<string, any>> = {
    request: {
      query: updateWidgetFocusMutation,
      variables: updateWidgetFocusVariables,
    },
    result: () => {
      updateWidgetFocusCalled = true;
      return { data: updateWidgetFocusErrorData };
    },
  };

  const mocks = [updateWidgetFocusErrorMock];
  const { baseElement } = render(
    <MockedProvider mocks={mocks}>
      <TextfieldPropertySection
        editingContextId="editingContextId"
        formId="formId"
        widget={defaultTextField}
        subscribers={[]}
        readOnly={false}
      />
    </MockedProvider>
  );
  expect(baseElement).toMatchSnapshot();

  const element: HTMLInputElement = screen.getByTestId('Name:').querySelector('input');
  expect(element.value).toBe('Composite Processor');

  await act(async () => {
    userEvent.click(element);
    await new Promise((resolve) => setTimeout(resolve, 0));

    await waitFor(() => {
      expect(updateWidgetFocusCalled).toBeTruthy();
      expect(screen.getByTestId('error')).not.toBeUndefined();
      expect(baseElement).toMatchSnapshot();
    });
  });
});

test('should render the textfield without style', () => {
  const { baseElement } = render(
    <MockedProvider>
      <TextfieldPropertySection
        editingContextId="editingContextId"
        formId="formId"
        widget={defaultTextField}
        subscribers={[]}
        readOnly={false}
      />
    </MockedProvider>
  );
  expect(baseElement).toMatchSnapshot();
});

test('should render the textfield with style', () => {
  const { baseElement } = render(
    <MockedProvider>
      <TextfieldPropertySection
        editingContextId="editingContextId"
        formId="formId"
        widget={textFieldWithStyle}
        subscribers={[]}
        readOnly={false}
      />
    </MockedProvider>
  );
  expect(baseElement).toMatchSnapshot();
});

test('should render the textfield with empty style', async () => {
  const { baseElement } = render(
    <MockedProvider>
      <TextfieldPropertySection
        editingContextId="editingContextId"
        formId="formId"
        widget={textFieldWithEmptyStyle}
        subscribers={[]}
        readOnly={false}
      />
    </MockedProvider>
  );
  expect(baseElement).toMatchSnapshot();
});
