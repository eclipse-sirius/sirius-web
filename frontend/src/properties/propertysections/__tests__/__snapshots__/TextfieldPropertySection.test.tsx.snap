// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should display the edited value 1`] = `
<div>
  <div>
    <div
      class="makeStyles-propertySectionLabel-13"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle2"
      >
        Name:
      </h6>
      <div
        class="makeStyles-subscribers-14"
      />
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
      data-testid="Name:"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
          name="Name:"
          placeholder="Name:"
          type="text"
          value="Composite Processor"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`should display the edited value 2`] = `
<div>
  <div>
    <div
      class="makeStyles-propertySectionLabel-13"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle2"
      >
        Name:
      </h6>
      <div
        class="makeStyles-subscribers-14"
      />
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
      data-testid="Name:"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth Mui-focused Mui-focused MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
          name="Name:"
          placeholder="Name:"
          type="text"
          value="Composite Processor"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`should display the edited value 3`] = `
<div>
  <div>
    <div
      class="makeStyles-propertySectionLabel-13"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle2"
      >
        Name:
      </h6>
      <div
        class="makeStyles-subscribers-14"
      />
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
      data-testid="Name:"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth Mui-focused Mui-focused MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
          name="Name:"
          placeholder="Name:"
          type="text"
          value="Main Composite Processor"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`should display the error received 1`] = `
<body>
  <div>
    <div>
      <div
        class="makeStyles-propertySectionLabel-18"
      >
        <h6
          class="MuiTypography-root MuiTypography-subtitle2"
        >
          Name:
        </h6>
        <div
          class="makeStyles-subscribers-19"
        />
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
        data-testid="Name:"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
            name="Name:"
            placeholder="Name:"
            type="text"
            value="Composite Processor"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`should display the error received 2`] = `
<body>
  <div>
    <div>
      <div
        class="makeStyles-propertySectionLabel-18"
      >
        <h6
          class="MuiTypography-root MuiTypography-subtitle2"
        >
          Name:
        </h6>
        <div
          class="makeStyles-subscribers-19"
        />
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
        data-testid="Name:"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth Mui-focused Mui-focused MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
            name="Name:"
            placeholder="Name:"
            type="text"
            value="Composite Processor"
          />
        </div>
      </div>
      <div
        class="MuiSnackbar-root MuiSnackbar-anchorOriginBottomRight"
        data-testid="error"
      >
        <div
          class="MuiPaper-root MuiSnackbarContent-root MuiPaper-elevation6"
          direction="up"
          role="alert"
          style="opacity: 1; transform: scale(1, 1); transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms,transform 150ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
        >
          <div
            class="MuiSnackbarContent-message"
          >
            An error has occurred, please refresh the page
          </div>
          <div
            class="MuiSnackbarContent-action"
          >
            <button
              aria-label="close"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-colorInherit MuiIconButton-sizeSmall"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`should render a readOnly textfield 1`] = `
<div>
  <div>
    <div
      class="makeStyles-propertySectionLabel-8"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle2"
      >
        Name:
      </h6>
      <div
        class="makeStyles-subscribers-9"
      />
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
      data-testid="Name:"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline Mui-disabled Mui-disabled MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input Mui-disabled Mui-disabled MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
          disabled=""
          name="Name:"
          placeholder="Name:"
          type="text"
          value="Composite Processor"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`should render the textfield 1`] = `
<div>
  <div>
    <div
      class="makeStyles-propertySectionLabel-3"
    >
      <h6
        class="MuiTypography-root MuiTypography-subtitle2"
      >
        Name:
      </h6>
      <div
        class="makeStyles-subscribers-4"
      />
    </div>
    <div
      class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
      data-testid="Name:"
    >
      <div
        class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
      >
        <input
          aria-invalid="false"
          class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
          name="Name:"
          placeholder="Name:"
          type="text"
          value="Composite Processor"
        />
      </div>
    </div>
  </div>
</div>
`;

exports[`should render the textfield with empty style 1`] = `
<body>
  <div>
    <div>
      <div
        class="makeStyles-propertySectionLabel-33"
      >
        <h6
          class="MuiTypography-root MuiTypography-subtitle2"
        >
          Name:
        </h6>
        <div
          class="makeStyles-subscribers-34"
        />
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
        data-testid="Name:"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-style-31 makeStyles-style-32 MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
            name="Name:"
            placeholder="Name:"
            type="text"
            value="Composite Processor"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`should render the textfield with style 1`] = `
<body>
  <div>
    <div>
      <div
        class="makeStyles-propertySectionLabel-28"
      >
        <h6
          class="MuiTypography-root MuiTypography-subtitle2"
        >
          Name:
        </h6>
        <div
          class="makeStyles-subscribers-29"
        />
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
        data-testid="Name:"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline makeStyles-style-26 makeStyles-style-27 MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
            name="Name:"
            placeholder="Name:"
            type="text"
            value="Composite Processor"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`should render the textfield without style 1`] = `
<body>
  <div>
    <div>
      <div
        class="makeStyles-propertySectionLabel-23"
      >
        <h6
          class="MuiTypography-root MuiTypography-subtitle2"
        >
          Name:
        </h6>
        <div
          class="makeStyles-subscribers-24"
        />
      </div>
      <div
        class="MuiFormControl-root MuiTextField-root MuiFormControl-marginDense MuiFormControl-fullWidth"
        data-testid="Name:"
      >
        <div
          class="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-fullWidth MuiInput-fullWidth MuiInputBase-formControl MuiInput-formControl MuiInputBase-marginDense MuiInput-marginDense"
        >
          <input
            aria-invalid="false"
            class="MuiInputBase-input MuiInput-input MuiInputBase-inputMarginDense MuiInput-inputMarginDense"
            name="Name:"
            placeholder="Name:"
            type="text"
            value="Composite Processor"
          />
        </div>
      </div>
    </div>
  </div>
</body>
`;
